feat: optionally position scrollbar over menu

willothy opened the pull request:
I prefer a floating scrollbar like I had originally implemented - this adds a "floating" option to use no background window, and an option to disable the scrollbar entirely. The current behavior is still the default.

willothy commented:
Rebased on top of master and updated docs :)

Bekaboo commented:
Hi, thanks for this PR. We could achieve further customizability by allowing user to control the window configurations of the scrollbar and enable/disable the scrollbar thumb and the background separately, like this:

```lua
local opts = {
  menu = {
    scrollbar = {
      enable = true,
        sbar = {
          enable = true,
          ---@alias dropbar_menu_win_config_opts_t any|fun(menu: dropbar_menu_t):any
          ---@type table<string, dropbar_menu_win_config_opts_t>
          ---@see vim.api.nvim_open_win
          win_conifgs = {
          }
        },
        thumb = {
          enable = true,
          ---@type table<string, dropbar_menu_win_config_opts_t>
          win_conifgs = {
          }
        }
      }
    }
  }
}
```

willothy commented:
That's a good idea, I'll make those changes tomorrow!

willothy commented:
My bad, some things came up and I didn't have time to get to this, but I should have some time in the next few days :)

Bekaboo commented:
@willothy Sure, please take your time, we are not in a hurry! Thanks for the great PRs you've made along the way. I can give you push permission if that is more convenient for you.

willothy commented:
@Bekaboo sure, that'd be cool! Thanks for trusting me with that :) I'll still wait for reviews before merging anything of course.

Bekaboo commented:
I have sent you an invitation.

willothy commented:
Thanks, sounds good!

willothy commented:
Working on this right now - I think it would be cool to be able to disable the bar and thumb separately, but I'm not sure I see a point in allowing users to customize `win_configs`. The position of the scrollbar should always be to the right of the menu, so it would likely end up ignoring all options except `border`, which I doubt people will use for a scrollbar. What do you think?

I think something like this is probably enough, and more properties could be added individually (maybe `blend` or `winhl`) if needed.

```lua
scrollbar = {
  enabled = true,
  -- The background / gutter of the scrollbar (optional)
  sbar = {
    -- When false, only the thumb is shown.
    enabled = true,
  },
},
```

Bekaboo commented:
> Also, the enable field should have type fun(menu: dropbar_menu_t)|boolean. The default value of sbar.enable can be a function that returns true if the border of menu is none else false, to avoid covering the right border of the menu, like how nvim-cmp's scrollbar works.

I think the reason why nvim-cmp's scrollbar has no background is simply that the background has the same `zindex` as the completion menu, so it is covered by the border of completion menu (if any), but the thumb has a higher `zindex` and is thus not covered.

willothy commented:
> `zindex` issue should be fixed in [5ace799](https://github.com/Bekaboo/dropbar.nvim/commit/5ace79950242b963c552760c6d050140a8676183), the scrollbar thumb will have a higher zindex than the menu while the sbar (background) will have the same zindex so that it will be converted by the menu border, if any.
> 
> ![Screenshot_20231203_125306](https://private-user-images.githubusercontent.com/76579810/287498326-ddf16a79-9ae7-47d3-9646-aef2d56fb100.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE1ODA1OTQsIm5iZiI6MTcwMTU4MDI5NCwicGF0aCI6Ii83NjU3OTgxMC8yODc0OTgzMjYtZGRmMTZhNzktOWFlNy00N2QzLTk2NDYtYWVmMmQ1NmZiMTAwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAzVDA1MTEzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiMDFlZjEyNTI5ZDA1NzFjN2EzZDNkMDhjZWNlMzE4YmY1MDg4MzYwNDE5ZjdlMjliYjNlOTRkOWI4ZmNjMjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0MDfdaL-K_mnv-knq-fmgC6QSMEMKfhr3lCND-xmNxY)

Nice, that looks really cool! I may have to start using a border in my config...

> > Also, the enable field should have type `fun(menu: dropbar_menu_t)|boolean`. The default value of sbar.enable can be a function that returns true if the border of menu is none else false, to avoid covering the right border of the menu, like how nvim-cmp's scrollbar works.
> 
> To avoid complexity the `enable` field should have type `boolean`, we will not allow a callback. @willothy What is your opinion?

I agree with the enable field just being a boolean, I think the border logic you just added is all that's needed since that's probably what people would use a callback for.

> This makes sense, to reduce the complexity we can only make the border configurable. However we can still have the win_configs field but the only valid sub-field is border, in case we want to make other sub-fields configurable in the future. This must be clarified in the README and vimdoc to avoid confusion.

That sounds good. Should the border be allowed on the background as well, or just on the thumb? There may be some edge cases to handle to ensure that scrollbar borders don't overlap with submenus.

Bekaboo commented:
> That sounds good. Should the border be allowed on the background as well, or just on the thumb? There may be some edge cases to handle to ensure that scrollbar borders don't overlap with submenus.

After some reflection, I realized that no one has really requested to customize the scrollbar border, and only allowing `border` to be configurable in the `win_configs` table is wired. If you need to customize it you can add it, if not I suggest drop this, so the config spec will look like the following:

```lua
scrollbar = {
  enabled = true,
  -- The background / gutter of the scrollbar (optional)
  sbar = {
    enabled = true, -- should we remove this too? it is still needed if you want to remove the background when menu border is 'none'
  },
},
```

willothy commented:
> After some reflection, I realized that no one has really requested to customize the scrollbar border, and only allowing `border` to be configurable in the `win_configs` table is wired.

Yeah, that's good with me, makes things a lot simpler. I wouldn't be using the border, so no worries. Not sure why I had that idea lol. 

> ...
> sbar = {
>   enabled = true, -- should we remove this too? it is still needed if you want to remove the background when menu border is 'none'
> },
> ...
> ```

I do want to still be able to remove the background when there's no border though, I think that's an important customization (or one that I'll use, at least haha). Maybe it can just be a single boolean, like `sbar = true` or `background = true` instead of a table?

willothy commented:
Should be all good now

Bekaboo commented:
@willothy Thanks for the PR, looks great!

A few small suggestions:

1. Use `enable` not `enabled`, for consistency with `general.enable`
2. New options should be documented under README [Menu](https://github.com/Bekaboo/dropbar.nvim#menu) section
3. Any reason to get the width of the menu window using `nvim_win_get_width()` instead of getting it from `menu_win_configs.width` on line 573 in `menu.lua`?
4. After 2b7c2d53363cb3d93376904dac3ea6d52dd900c5, the background (sbar) is first created, then the thumb. Could you also reorder it in the PR? Sorry for the inconvenience.

willothy commented:
> 1. Use `enable` not `enabled`, for consistency with `general.enable`

Done

> 2. New options should be documented under README [Menu](https://github.com/Bekaboo/dropbar.nvim#menu) section and also in the vimdoc

Added vimdoc and updated the scrollbar docs in readme with the new names.

> 3. Any reason to get the width of the menu window using `nvim_win_get_width()` instead of getting it from `menu_win_configs.width` on line 573 in `menu.lua`?

Not sure, but I removed it because it's not needed now lol

> 4. After [2b7c2d5](https://github.com/Bekaboo/dropbar.nvim/commit/2b7c2d53363cb3d93376904dac3ea6d52dd900c5), the background (sbar) is created first, then the thumb. Could you also reorder it in the PR? Sorry for the inconvenience.

No worries! Should be done as well.
